boat_simulator/nodes/physics_engine/kinematics_computation.py

```diff
@@ -0,0 +1,125 @@
+    __global_data = KinematicsData()
+
+    def __init__(self, timestep: Scalar, mass: Scalar, inertia: ArrayLike) -> None:
+        self.__timestep = timestep
+        self.__boat_mass = mass
+        self.__inertia = np.array(inertia, dtype=np.float32)
+        self.__inertia_inverse = np.linalg.inv(self.__inertia)
+
+    # Updates the kinematics data and returns the kinematics data objects
+    def step(self, net_force: ArrayLike, net_torque: ArrayLike
+             ) -> Tuple[KinematicsData, KinematicsData]:
+
+        next_relative_acceleration, next_relative_velocity = \
+            self.__update_linear_relative_data(net_force)
+
+        yaw_radians = self.__update_ang_data(net_torque)
+
+        self.__update_linear_global_data(
+            next_relative_acceleration, next_relative_velocity, yaw_radians
+            )
+
+        return (self.relative_data, self.global_data)
+
+    def __update_linear_relative_data(self, net_force: ArrayLike) -> Tuple[ArrayLike, ArrayLike]:
+        next_relative_acceleration = self.__compute_next_relative_acceleration(
+            self.boat_mass, net_force
+            )
+        next_relative_velocity = self.__compute_next_velocity(
+            self.relative_data.linear_velocity, self.relative_data.linear_acceleration
+            )
+
+        self.__relative_data.linear_acceleration = next_relative_acceleration
+        self.__relative_data.linear_velocity = next_relative_velocity
+
+        return (next_relative_acceleration, next_relative_velocity)
+
+    def __update_linear_global_data(self, next_relative_acceleration: ArrayLike,
+                                    next_relative_velocity: ArrayLike,
+                                    yaw_radians: Scalar) -> None:
+
+        next_global_acceleration = next_relative_acceleration * \
+            np.array([np.cos(yaw_radians), np.sin(yaw_radians), 0])
+        next_global_velocity = next_relative_velocity * \
+            np.array([np.cos(yaw_radians), np.sin(yaw_radians), 0])
+        next_global_position = self.__compute_next_position(
+            self.global_data.linear_position,
+            self.global_data.linear_velocity,
+            self.global_data.linear_acceleration
+            )
+
+        self.__global_data.linear_acceleration = next_global_acceleration
+        self.__global_data.linear_velocity = next_global_velocity
+        self.__global_data.linear_position = next_global_position
+
+    def __update_ang_data(self, net_torque: ArrayLike) -> Scalar:
+        next_ang_acceleration = self.__compute_next_ang_acceleration(net_torque)
+        next_ang_velocity = self.__compute_next_velocity(
+            self.relative_data.angular_velocity, self.relative_data.angular_acceleration
+            )
+        next_ang_position = self.__compute_next_position(
+            self.relative_data.angular_position,
+            self.relative_data.angular_velocity,
+            self.relative_data.angular_acceleration
+            )
```

Comment:
Remember to bound all angles between $[-\pi, \pi)$ radians. Could you implement the `bound_to_180` function in `common/utils.py` and use it here? It says it accepts scalars right now, but change to `ArrayLike`.

For the most efficient solution, look at the Sailbot software quiz solution (it's a private repo, but you should be able to see it).

Reply:
You probably want to use the global data to update and not the relative data to update the position.